File: vol1/ch3/LeetCode496.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        dict={}
        m,n=len(nums1),len(nums2)
        st=[]              #小顶单调栈
        for i in range(0,n):
            while len(st)>0 and st[-1]<nums2[i]:
                dict[st[-1]]=nums2[i]
                st.pop()           #将小于nums[i]的元素出栈
            st.append(nums2[i])
        ans=[-1]*m
        for i in range(0,m):
            if dict.get(nums1[i])==None:
                continue;
            ans[i]=dict.get(nums1[i])
        return ans

File: vol1/ch3/LeetCode496.cpp
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        unordered_map<int,int> hmap;
        int m=nums1.size();
        int n=nums2.size();
        stack<int> st;              //小顶栈
        for (int i=0;i<n;i++) {
            while (!st.empty() && st.top()<nums2[i]) {
                hmap[st.top()]=nums2[i];
                st.pop();           //将小于nums[i]的元素出栈
            }
            st.push(nums2[i]);
        }
        vector<int> ans(m,-1);					//初始化所有元素为-1
        for (int i=0;i<m;i++) {
            if (hmap.find(nums1[i])==hmap.end())
                continue;
            ans[i]=hmap[nums1[i]];
        }
        return ans;
    }
};

